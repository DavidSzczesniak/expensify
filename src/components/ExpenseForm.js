import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: '',
        };
    }
    render() {
        const { description, note, amount, createdAt, calendarFocused, error } = this.state;

        const handleDescriptionChange = (description) => {
            this.setState(() => ({ description }));
        };

        const handleNoteChange = (note) => {
            this.setState(() => ({ note }));
        };

        const handleDateChange = (createdAt) => {
            createdAt && this.setState(() => ({ createdAt }));
        };

        const handleFocusChange = ({ focused }) => {
            this.setState(() => ({ calendarFocused: focused }));
        };

        const handleAmountChange = (amount) => {
            if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
                this.setState(() => ({ amount }));
            }
        };

        const handleSubmit = (e) => {
            e.preventDefault();
            if (!description || !amount) {
                this.setState(() => ({ error: 'Please provide a description and amount' }));
            } else {
                this.setState(() => ({ error: '' }));
                this.props.onSubmit({
                    description,
                    amount: parseFloat(amount, 10) * 100,
                    createdAt: createdAt.valueOf(),
                    note,
                });
            }
        };

        return (
            <div>
                {error && <p>{error}</p>}
                <form onSubmit={handleSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={description}
                        onChange={(e) => handleDescriptionChange(e.target.value)}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={amount}
                        onChange={(e) => handleAmountChange(e.target.value)}
                    />
                    <SingleDatePicker
                        date={createdAt}
                        onDateChange={handleDateChange}
                        focused={calendarFocused}
                        onFocusChange={handleFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note (optional)"
                        value={note}
                        onChange={(e) => handleNoteChange(e.target.value)}
                    />
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}
