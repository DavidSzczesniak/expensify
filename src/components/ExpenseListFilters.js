import React from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux';
import {
    setEndDate,
    setStartDate,
    setTextFilter,
    sortByAmount,
    sortByDate,
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    render() {
        const { filters, setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } =
            this.props;

        const handleTextChange = (e) => {
            setTextFilter(e.target.value);
        };

        const handleFilterSelect = (value) => {
            if (value === 'amount') {
                sortByAmount();
            } else if (value === 'date') {
                sortByDate();
            }
        };

        const handleDatesChange = ({ startDate, endDate }) => {
            setStartDate(startDate);
            setEndDate(endDate);
        };

        const handleFocusChange = (calendarFocused) => {
            this.setState(() => ({ calendarFocused }));
        };

        return (
            <div>
                <input type="text" value={filters.text} onChange={handleTextChange} />
                <select value={filters.sortBy} onChange={(e) => handleFilterSelect(e.target.value)}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={filters.startDate}
                    endDate={filters.endDate}
                    onDatesChange={handleDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={handleFocusChange}
                    showClearDates
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
