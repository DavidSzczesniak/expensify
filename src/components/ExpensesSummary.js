import React from 'react';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import { connect } from 'react-redux';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const countWord = expenseCount === 1 ? 'expense' : 'expenses';
    const total = numeral(expensesTotal / 100).format('$0,0.00');

    return (
        <div>
            <p>
                Viewing {expenseCount} {countWord} totalling {total}
            </p>
        </div>
    );
};

const mapStateToProps = (state) => {
    const selectedExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: selectedExpenses.length,
        expensesTotal: getExpensesTotal(selectedExpenses),
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
