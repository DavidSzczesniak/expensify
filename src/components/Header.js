import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <div>
        <h1>Expensify</h1>
        <NavLink exact to="/">
            Home
        </NavLink>
        <NavLink to="/create">Create Expense</NavLink>
        <NavLink to="/help">Help</NavLink>
    </div>
);

export default Header;
