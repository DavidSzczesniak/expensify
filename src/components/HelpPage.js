import React from 'react';

const HelpPage = () => <div>Help page here</div>;

export default HelpPage;
