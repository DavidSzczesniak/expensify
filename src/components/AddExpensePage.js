import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export const AddExpensePage = ({ startAddExpense, history }) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm
            onSubmit={(expense) => {
                startAddExpense(expense);
                history.push('/');
            }}
        />
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
