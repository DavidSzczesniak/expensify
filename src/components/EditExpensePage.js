import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export const EditExpensePage = ({ expense, history, editExpense, removeExpense }) => (
    <div>
        <ExpenseForm
            onSubmit={(editedExpense) => {
                editExpense(expense.id, editedExpense);
                history.push('/');
            }}
            expense={expense}
        />
        <button
            onClick={() => {
                removeExpense({ id: expense.id });
                history.push('/');
            }}>
            Remove
        </button>
    </div>
);

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
