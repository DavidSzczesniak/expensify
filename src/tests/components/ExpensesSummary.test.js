import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render component correctly with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={9434} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render component correctly with 2 expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={12343} />);
    expect(wrapper).toMatchSnapshot();
});
