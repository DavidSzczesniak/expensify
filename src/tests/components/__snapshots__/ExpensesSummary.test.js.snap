// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $94.34
  </p>
</div>
`;

exports[`should render component correctly with 2 expenses 1`] = `
<div>
  <p>
    Viewing 
    2
     
    expenses
     totalling 
    $123.43
  </p>
</div>
`;
